### PR TITLE
Add ThrowOnPartialError to Azure.Monitor.Query

### PR DESCRIPTION
**Pull request by pakrym**
Add partial failure plan to LogsQueryResult
Add Collection type to batch results

https://github.com/Azure/azure-sdk-for-net/issues/23580

### CODE REVIEW

File: sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultCollection.cs [hunk 1]
```diff
@@ -33,32 +36,39 @@ public partial class LogsBatchQueryResults
...
         /// </code>
         /// <param name="queryId">The query identifier returned from the <see cref="LogsBatchQuery.AddQuery"/>.</param>
-        /// <returns>The <see cref="LogsBatchQueryResults"/> with the query results.</returns>
+        /// <returns>The <see cref="LogsBatchQueryResultCollection"/> with the query results.</returns>
         /// <exception cref="ArgumentException">When the query with <paramref name="queryId"/> was not part of the batch.</exception>
         /// <exception cref="RequestFailedException">When the query <paramref name="queryId"/> failed.</exception>
         public LogsBatchQueryResult GetResult(string queryId)
         {
-            BatchQueryResponse result = Responses.SingleOrDefault(r => r.Id == queryId);
+            LogsBatchQueryResult result = this.SingleOrDefault(r => r.Id == queryId);
```
Comment (christothes) [inline sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultCollection.cs:50]: Not sure if the description or the logic is more correct, but going by the description, I'd expect this to be an `Any` if the description is accurate

File: sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultStatus.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    /// Represents a status of a batch query response.
+    /// </summary>
+    public enum LogsBatchQueryResultStatus
+    {
+        /// <summary>
+        /// The query succeeded.
+        /// </summary>
+        Success,
+        /// <summary>
+        /// The query partially succeeded, both data and an error was returned.
```
Comment (christothes) [inline sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultStatus.cs:16]: "at least one error"?

File: sdk/monitor/Azure.Monitor.Query/src/Models/LogsQueryResults.cs [hunk 3]
```diff
@@ -57,5 +57,18 @@ public LogsTable Table
...
         public ResponseError Error => _error.ValueKind == JsonValueKind.Undefined ? null : JsonSerializer.Deserialize<ResponseError>(_error.GetRawText());
+
+        internal Exception CreateExceptionForErrorResponse(int status)
+        {
+            return new RequestFailedException(
+                status,
+                $"The result was returned but contained a partial error, you can disable an exception for " +
+                $"partial errors using {nameof(LogsQueryOptions)}.{nameof(LogsQueryOptions.ThrowOnPartialErrors)} " +
+                $"and inspect the partial error using {nameof(LogsQueryResult)}.{nameof(Error)} property.{Environment.NewLine}" +
+                $"Error:{Environment.NewLine}{Error}",
```
Comment (christothes) [inline sdk/monitor/Azure.Monitor.Query/src/Models/LogsQueryResults.cs:68]:
```suggestion
                $"The result was returned but contained a partial error. Exceptions for partial errors can be disabled " +
                $" using {nameof(LogsQueryOptions)}.{nameof(LogsQueryOptions.ThrowOnPartialErrors)}." +
                $"Partial errors can be inspected using the {nameof(LogsQueryResult)}.{nameof(Error)} property.{Environment.NewLine}" +
                $"Error:{Environment.NewLine}{Error}",
```

**Review by christothes — APPROVED** [review]

File: sdk/monitor/Azure.Monitor.Query/src/Models/BatchQueryResponse.cs [hunk 4]
```diff
@@ -9,13 +9,18 @@ internal partial class BatchQueryResponse
     {
         /// <summary> Contains the tables, columns &amp; rows resulting from the query or the error details if the query failed. </summary>
         [CodeGenMember("Body")]
-        private LogsBatchQueryResult _body;
+        private readonly LogsBatchQueryResult _body;
 
         public LogsBatchQueryResult Body
```
[the remark this answers is not in the document]
Comment (pakrym) [reply sdk/monitor/Azure.Monitor.Query/src/Models/BatchQueryResponse.cs]: deleted the entire thing

File: sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultStatus.cs (hunk 2 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultStatus.cs:16]: This one is for an individual response in a batch. For log analytics, every request can partially fail.

File: sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultCollection.cs (hunk 1 above)
Comment (pakrym) [reply sdk/monitor/Azure.Monitor.Query/src/Models/LogsBatchQueryResultCollection.cs:50]: Talked offline.

**Pull request merged**